crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2576,7 +2609,53 @@ impl<'db> TypeInferenceBuilder<'db> {
                         Type::Unknown
                     })
             }
-            _ => Type::Todo,
+            (value_ty, slice_ty) => {
+                // Resolve the value to its class.
+                let value_meta_ty = value_ty.to_meta_type(self.db);
+
+                // If the class defines `__getitem__`, return its return type.
+                //
+                // See: https://docs.python.org/3/reference/datamodel.html#class-getitem-versus-getitem
+                let dunder_getitem_method = value_meta_ty.member(self.db, "__getitem__");
+                if !dunder_getitem_method.is_unbound() {
+                    // TODO validate that the `__getitem__` argument adheres to the signature
+                    let CallOutcome::Callable { return_ty } =
+                        dunder_getitem_method.call(self.db, &[slice_ty])
+                    else {
+                        self.dunder_not_callable_diagnostic(
```

Comment:
As @AlexWaygood mentioned in another comment, matching directly on `CallOutcome` this way doesn't work correctly for calling unions. It also bypasses the diagnostics emitted by `CallOutcome::unwrap_with_diagnostic`, meaning the user gets no feedback about what the type of `__getitem__` is, that isn't callable.

The intended way to use `CallOutcome` is to just call `unwrap_with_diagnostic` on it, and not necessarily emit your own diagnostic on top of that. In this case, that means the user won't get the context that it's the `__getitem__` that wasn't callable, so we may want to add a way to pass additional context into `CallOutcome::unwrap_with_diagnostic` further identifying what it is that we tried and failed to call. (This would then mean we don't need `dunder_not_callable_diagnostic` anymore.)